src/GraphQL/Validation/Rules/NoFragmentCycles.cs

```diff
@@ -20,36 +20,28 @@ public class NoFragmentCycles : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="NoFragmentCyclesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            // Tracks already visited fragments to maintain O(N) and to ensure that cycles
-            // are not redundantly reported.
-            var visitedFrags = new LightweightCache<string, bool>(key => false);
-
-            // Array of AST nodes used to produce meaningful errors
-            var spreadPath = new Stack<FragmentSpread>();
-
-            // Position in the spread path
-            var spreadPathIndexByName = new LightweightCache<string, int>(key => -1);
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => context.Document.Fragments.Count > 0 ? _nodeVisitor : null;
 
-            return new MatchingNodeVisitor<FragmentDefinition>(node =>
-                {
-                    if (!visitedFrags[node.Name])
-                    {
-                        detectCycleRecursive(node, spreadPath, visitedFrags, spreadPathIndexByName, context);
-                    }
-                }).ToTask();
-        }
+        private static readonly Task<INodeVisitor> _nodeVisitor = new MatchingNodeVisitor<FragmentDefinition>((node, context) =>
+        {
+            var visitedFrags = context.TypeInfo.NoFragmentCycles_VisitedFrags ??= new HashSet<string>();
```

Comment:
This was a dictionary, but the value was not used.  Changed to a HashSet

Reply:
Actually, even worse, it was a LightweightCache.

Reply:
Note that we could easily add an extension method `GetOrDefault` to a dictionary to perform the same effect as LightweightCache.  I initially did so, but there was so little need of it that I removed it.